kv: get rid of orderingPolicy telling the transport how to order

andreimatei opened the pull request:
replicas

It would appear that we were shuffling replicas twice before sending
RPCs to them. Once in distSender.optimizeReplicaOrder, and then again in
grpcTransport. Removed the second one. Also optimizeReplicaOrder was
confusing because it both applied an order, and returned an ordering
policy to the caller. I've removed the ordering policy and commented on
the semantics of optimizeReplicaOrder.

I'm doing this because I want to lift optimizeReplicaOrder and got
confused about the semantics.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6807)

<!-- Reviewable:end -->

andreimatei committed:
kv: get rid of orderingPolicy telling the transport how to order
replicas

It would appear that we were shuffling replicas twice before sending
RPCs to them. Once in distSender.optimizeReplicaOrder, and then again in
grpcTransport. Removed the second one. Also optimizeReplicaOrder was
confusing because it both applied an order, and returned an ordering
policy to the caller. I've removed the ordering policy and commented on
the semantics of optimizeReplicaOrder.

I'm doing this because I want to lift optimizeReplicaOrder and got
confused about the semantics.

andreimatei commented:
cc @tschottdorf 

I don't know what I'm doing, but I hear that's why we have code reviews.

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> #### kv: get rid of orderingPolicy telling the transport how to order
> 
> replicas
> 
> 
> 
> It would appear that we were shuffling replicas twice before sending
> 
> RPCs to them. Once in distSender.optimizeReplicaOrder, and then again in
> 
> grpcTransport. Removed the second one. Also optimizeReplicaOrder was
> 
> confusing because it both applied an order, and returned an ordering
> 
> policy to the caller. I've removed the ordering policy and commented on
> 
> the semantics of optimizeReplicaOrder.
> 
> 
> 
> I'm doing this because I want to lift optimizeReplicaOrder and got
> 
> confused about the semantics.
> 
> 
> 
</details>


---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6807#-:-KI9QCAfT7pch77kReWG:1488056198)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

For historical context, I'm pretty sure this structure was intended to avoid the cost of shuffling the list if we could come up with any reason to switch to stable mode. But when `SortByCommonAttributePrefix` was changed to break ties randomly (in acca40e9d), we shuffled early anyway and it became redundant. 

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> cc @tschottdorf 
> 
> I don't know what I'm doing, but I hear that's why we have code reviews.
</details>


---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[kv/dist_sender.go, line 262 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6807#-KIAD1bwO8rZgCaSM7Dl:-KIAD1bwO8rZgCaSM7Dm:-429072005) ([raw file](https://github.com/cockroachdb/cockroach/blob/207b145f7bbdea43f25d94f1c5632236905474ed/kv/dist_sender.go#L262)):*
> ```Go
> // If the current node is a replica, then it'll be the first one.
> //
> // nodeDesc is the descriptor of the current node. It can be nil, in which case
> ```

This part of the comment looks out of place. Were you trying to make `nodeDesc` an argument? 

---

*[kv/dist_sender.go, line 265 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6807#-KIAEM8h5QEzrEBBRA-p:-KIAEM8h5QEzrEBBRA-q:-1774492247) ([raw file](https://github.com/cockroachdb/cockroach/blob/207b145f7bbdea43f25d94f1c5632236905474ed/kv/dist_sender.go#L265)):*
> ```Go
> // information about the current descriptor is not used in optimizing the order.
> func (ds *DistSender) optimizeReplicaOrder(replicas ReplicaSlice) {
> 	// TODO(spencer): going to need to also sort by affinity; closest
> ```

FYI when we get around to doing this it will probably have to live on the transport, so feel free to refactor this to move `optimizeReplicaOrder` closer to the transport if it helps with what you're doing.

---

*[kv/transport.go, line 133 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6807#-KIAARt7oiHf8ESOakeR:-KIAARt8VdqsWo9jQJAw:-453797591) ([raw file](https://github.com/cockroachdb/cockroach/blob/207b145f7bbdea43f25d94f1c5632236905474ed/kv/transport.go#L133)):*
> ```Go
> 
> 	// Put known-unhealthy clients last.
> 	_, err := splitHealthy(clients)
> ```

This is the only use of `splitHealthy`. If we don't need the first return value here, we don't need it anywhere so we can just change the type of the function. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6807#-:-KIAEYbroriNu0nLUpM8:-2141486729)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[kv/dist_sender.go, line 262 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6807#-KIAD1bwO8rZgCaSM7Dl:-KIAmRsNVoiVH50zId3R:-1621224908) ([raw file](https://github.com/cockroachdb/cockroach/blob/207b145f7bbdea43f25d94f1c5632236905474ed/kv/dist_sender.go#L262)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This part of the comment looks out of place. Were you trying to make `nodeDesc` an argument? 
</details>
ha, yes, I split this commit out of another branch where this function doesn't sit in the `DistSender` anymore. Fixed (for now).

---

*[kv/dist_sender.go, line 265 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6807#-KIAEM8h5QEzrEBBRA-p:-KIAn4y75q2REh85SLjg:-477773138) ([raw file](https://github.com/cockroachdb/cockroach/blob/207b145f7bbdea43f25d94f1c5632236905474ed/kv/dist_sender.go#L265)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> FYI when we get around to doing this it will probably have to live on the transport, so feel free to refactor this to move `optimizeReplicaOrder` closer to the transport if it helps with what you're doing.
</details>
well, it actually helps me to move it away from the transport. Since I wanna use it in the leader-cache-wrapper that will use this function for guessing/electing a leader.
So if transport starts keeping track of latencies, I'd like figure out a way to expose it to other layers.

---

*[kv/transport.go, line 133 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6807#-KIAARt7oiHf8ESOakeR:-KIAo2z1fGzKhteBozB7:489589776) ([raw file](https://github.com/cockroachdb/cockroach/blob/207b145f7bbdea43f25d94f1c5632236905474ed/kv/transport.go#L133)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is the only use of `splitHealthy`. If we don't need the first return value here, we don't need it anywhere so we can just change the type of the function. 
</details>
I noticed that, but it seems a reasonable thing for `splitHealthy` to return and it doesn't cost the implementation of that function any code. I'd leave it, even if unused...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6807)*
<!-- Sent from Reviewable.io -->
